x-pack/plugins/observability_solution/infra/server/routes/services/index.ts

```diff
@@ -33,21 +32,39 @@ export const initServicesRoute = (libs: InfraBackendLibs) => {
         },
       },
     },
-    async (requestContext, request, response) => {
-      const [{ savedObjects }] = await libs.getStartServices();
+    async (context, request, response) => {
       const { from, to, size = 10, validatedFilters } = request.query;
 
-      const client = createSearchClient(requestContext, framework, request);
-      const soClient = savedObjects.getScopedClient(request);
-      const apmIndices = await libs.plugins.apmDataAccess.setup.getApmIndices(soClient);
-      const services = await getServices(client, apmIndices, {
-        from,
-        to,
-        size,
-        filters: validatedFilters!,
+      const apmDataAccessClient = getApmDataAccessClient({ request, libs, context });
+      const hasApmPrivileges = await apmDataAccessClient.hasPrivileges();
+
+      const apmDataAccessServices = hasApmPrivileges
+        ? await apmDataAccessClient.getServices()
+        : undefined;
+
+      const apmDocumentSources = await apmDataAccessServices?.getDocumentSources({
+        start: from,
+        end: to,
       });
-      return response.ok({
-        body: ServicesAPIResponseRT.encode(services),
+
+      if (apmDataAccessServices && apmDocumentSources) {
+        const services = await apmDataAccessServices?.getHostServices({
+          documentSources: apmDocumentSources,
+          start: from,
+          end: to,
+          filters: validatedFilters!,
+          size,
+        });
+        return response.ok({
+          body: ServicesAPIResponseRT.encode(services),
+        });
+      }
+
+      return response.customError({
+        statusCode: 403,
+        body: {
+          message: 'APM data access service is not available',
+        },
```

Comment:
Perhaps this will be even better if we return this right after `hasApmPrivileges` returns `false`. It make the function easier to read.

```ts
if (!hasApmPrivilege) {
  return response.customError({
    statusCode: 403,
    body: {
      message: 'APM data access service is not available',
    }
}
```